contrib/node/src/python/pants/contrib/node/tasks/node_build.py

```diff
@@ -0,0 +1,89 @@
+# coding=utf-8
+# Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+from collections import defaultdict
+
+from pants.backend.jvm.tasks.classpath_products import ClasspathProducts
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.goal.products import MultipleRootedProducts
+from pants.util.contextutil import pushd
+from pants.util.dirutil import absolute_symlink
+
+from pants.contrib.node.tasks.node_paths import NodePaths
+from pants.contrib.node.tasks.node_task import NodeTask
+
+
+class NodeBuild(NodeTask):
+  """Create an archive bundle of NodeModule targets."""
+
+  @classmethod
+  def product_types(cls):
+    # runtime_classpath is used for JVM target to include node build results as resources.
+    return ['bundleable_js', 'runtime_classpath']
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(NodeBuild, cls).prepare(options, round_manager)
+    round_manager.require_data(NodePaths)
+
+  @property
+  def create_target_dirs(self):
+    return True
+
+  def __init__(self, *args, **kwargs):
+    super(NodeBuild, self).__init__(*args, **kwargs)
+
+  def execute(self):
+    node_paths = self.context.products.get_data(NodePaths)
+    runtime_classpath_product = self.context.products.get_data(
+      'runtime_classpath', init_func=ClasspathProducts.init_func(self.get_options().pants_workdir))
+    bundleable_js_product = self.context.products.get_data(
+      'bundleable_js', init_func=lambda: defaultdict(MultipleRootedProducts))
+
+    targets = self.context.targets(predicate=self.is_node_module)
+    with self.invalidated(targets) as invalidation_check:
+      for vt in invalidation_check.all_vts:
```

Comment:
When using an invalidated block, I strongly recommend making the block as small as possible, and breaking out the "thing to do only if a target is invalid" and "things to do regardless" bits as methods.

The goal is to make it easy to read the block itself, since errors there lead to cache bugs.